src/hotspot/share/code/compressedStream.cpp

```diff
@@ -112,3 +112,106 @@ void CompressedWriteStream::write_long(jlong value) {
   write_signed_int(low(value));
   write_signed_int(high(value));
 }
+
+
+bool CompressedSparseDataReadStream::read_zero() {
+  if (_buffer[_position] & (1 << (7 - _bit_pos))) {
+    return 0; // not a zero data
+  }
+  if (++_bit_pos == 8) {
+    _position++;
+    _bit_pos = 0;
+  }
+  return 1;
+}
+
+uint8_t CompressedSparseDataReadStream::read_byte_impl() {
+  uint8_t b1 = _buffer[_position] << _bit_pos;
+  uint8_t b2 = _buffer[++_position] >> (8 - _bit_pos);
+  return b1 | b2;
+}
+
+jint CompressedSparseDataReadStream::read_int() {
+  if (read_zero()) {
+    return 0;
+  }
+  // integer value encoded as a sequence of 1 to 5 bytes
+  // - the most frequent case (0 < x < 64) is encoded in one byte
+  // - the payload of the first byte is 6 bits, the payload of the following bytes is 7 bits
+  // - the most significant bit in the first byte is occupied by a zero flag
+  // - each byte has a bit indicating whether it is the last byte in the sequence
+  //
+  //       value | byte0    | byte1    | byte2    | byte3    | byte4
+  //  -----------+----------+----------+----------+----------+----------
+  //           0 | 0        |          |          |          |
+  //           1 | 10000001 |          |          |          |
+  //           2 | 10000010 |          |          |          |
+  //          63 | 10111111 |          |          |          |
+  //          64 | 11000000 | 00000001 |          |          |
+  //          65 | 11000001 | 00000001 |          |          |
+  //        8191 | 11111111 | 01111111 |          |          |
+  //        8192 | 11000000 | 10000000 | 00000001 |          |
+  //        8193 | 11000001 | 10000000 | 00000001 |          |
+  //     1048575 | 11111111 | 11111111 | 01111111 |          |
+  //     1048576 | 11000000 | 10000000 | 10000000 | 00000001 |
+  //  0xFFFFFFFF | 11111111 | 11111111 | 11111111 | 11111111 | 00011111
+  //
+  uint8_t b = read_byte_impl();
+  juint result = b & 0x3f;
+  for (int i = 0; (i == 0) ? (b & 0x40) : (b & 0x80); i++) {
+    b = read_byte_impl();
+    result |= ((b & 0x7f) << (6 + 7 * i));
+  }
+  return (jint)result;
+}
+
+int CompressedSparseDataWriteStream::position() {
+  if (_bit_pos == 0) {
+    return _position;
+  }
+  // flush current data and start a new byte
+  write(_curr_byte << (8 - _bit_pos));
+  _curr_byte = 0;
+  _bit_pos = 0;
+  return _position;
+}
+
+void CompressedSparseDataWriteStream::write_zero() {
```

Comment:
We can rewrite the function not to use `_curr_byte`. We work directly on `_buffer` and rename `_bit_pos` into `_used_bits`:
```
_buffer[_position] >>= 1;
if (++_used_bits == 8) {
  _position += 1;
  _used_bits = 0; 
}

Reply:
thanks for the idea not to use _curr_byte
_bit_pos is used in both Read and Write streams. I think proper name is _bit_position.